feature/coordinators

letammyh opened the pull request:
Utilities: reorganized files, created UIStoryboardExtensions, Coordinator protocol, ViewLifecycleObserver
UI, Root: created AppCoordinator
UI, Menu, created MenuCoordinator, added instance of lifecycleObserver and class func make() to Menu VC
UI, UserGrid: created UserGridCoordinator, added instance of lifecycleObserver and class func make() to UserGrid VC

Commit:
Utilities: reorganized files, created UIStoryboardExtensions, Coordinator protocol, ViewLifecycleObserver
UI, Root: created AppCoordinator
UI, Menu, created MenuCoordinator, added instance of lifecycleObserver and class func make() to Menu VC
UI, UserGrid: created UserGridCoordinator, added instance of lifecycleObserver and class func make() to UserGrid VC

Commit:
Implemented MenuCoordinator, MenuWorkflow, and updated the MenuViewController.
Added to the UserGridCoordinator (still bare minimum, yet to handle loading userRecords from store instead of in ViewController)

Nitewriter commented on NameGame/NameGame/AppDelegate.swift:
```diff
@@ -27,10 +28,22 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
+        }
+        appCoordinator = AppCoordinator(store: store, container: navigationController)
+        store.subscribe(logger)
         FilePath.createDirectories()
```
Let's call `createDirectories` before we set everything else up. Just in case something we set up depends on it.

Nitewriter commented on NameGame/NameGame/State/Middleware/StateLogger.swift:
```diff
@@ -11,7 +11,8 @@ import ReSwift
     
     func newState(state: AppState) {
-        print("GameState: \(String(describing: state.gameState))")
+//        print("GameState: \(String(describing: state.gameState))")print("GameState: \(String(describing: state.gameState))")
```
How about instead of having the game state logging commented out, we add a mode to our logger.

```
enum LogLevel {
  case fullState
  case gameState
}

var logLevel: LogLevel = .fullState

fun newState(state: AppState {
  switch logLevel {
    ...
  }
}
```

Nitewriter commented on NameGame/NameGame/UI/Menu/MenuCoordinator.swift:
```diff
@@ -0,0 +1,53 @@
+    
+    func createGameCoordinator() {
+        // TODO
+    }
```
Style; add a new line after this function.

Nitewriter commented on NameGame/NameGame/UI/Menu/MenuWorkflow.swift:
```diff
@@ -0,0 +1,61 @@
+import Foundation
+import ReSwift
+
+protocol CoordinatorHandler {
```
Let's change this protocol to a `: class` type. Also possibly rename to `MenuWorkflowPresentationEventObserver`.

Nitewriter commented on NameGame/NameGame/UI/Menu/MenuWorkflow.swift:
```diff
@@ -0,0 +1,61 @@
+    
+    let store: Store<AppState>
+    private(set) var userRecordService: UserRecordService!
+    var coordinatorHandler: CoordinatorHandler!
```
We will want this to `weak` and optional so we don't create a retain cycle leak.

Nitewriter replied:
Also, if we rename the protocol; rename this property to `presentationEventHandler`.

Nitewriter commented on NameGame/NameGame/UI/Menu/ViewController/MenuViewController.swift:
```diff
@@ -0,0 +1,54 @@
+    @IBOutlet weak var teamMembersButton: UIButton!
+    
+    private var eventHandler: MenuViewEventHandler!
+//    weak var lifecycleObserver: ViewLifecycleObserver? = nil
```
Let's remove this commented code until it is needed.

Nitewriter commented on NameGame/NameGame/UI/Menu/MenuCoordinator.swift:
```diff
@@ -0,0 +1,53 @@
+    
+    func createUserGridCoordinator() {
+        let coordinator = UserGridCoordinator(store: store, container: container)
+        coordinator.start()
```
Looks like we are not holding on to a reference of this coordinator. May want to add the `currentCoordinator` bit you have in the `AppCoordinator`.

Nitewriter commented on NameGame/NameGame/Utilities/ViewLifecycleObserver.swift:
```diff
@@ -0,0 +1,26 @@
+    func viewDidAppear(_ animated: Bool) {}
+    func viewWillDisappear(_ animated: Bool) {}
+    func viewDidDisappear(_ animated: Bool) {}
+
```
Style; remove extra new line here.

Nitewriter reviewed (changes requested):
So close, I would love to see it in action when you have a moment. Also, we could chat for a second on Handler vs EventObserver.

Commit:
AppCoordinator: initialized ImageCache here and injected to coordinators and UserGridVC
Completed implementation of UserGrid Coord, ViewModel, VC
Coordinator: ControllerEventObserver protocol created

Nitewriter commented on NameGame/NameGame/UI/Menu/MenuCoordinator.swift:
```diff
@@ -0,0 +1,75 @@
+extension MenuCoordinator: CoordinatorEventObserver {
+    
+    func willStop(coordinator: Coordinator) {
+        if ((coordinator as? UserGridCoordinator) != nil) {
```
I would remove this conditional check for now until we need to do something specific with the coordinator on tear down.

Nitewriter commented on NameGame/NameGame/UI/Menu/MenuWorkflow.swift:
```diff
@@ -0,0 +1,59 @@
+import ReSwift
+
+protocol MenuWorkflowPresentationEventObserver: class {
+    func createUserGridCoordinator()
```
Maybe swap the word `create` with `present` for both of these interfaces.

Nitewriter commented on NameGame/NameGame/UI/UserGrid/UserGridViewModel.swift:
```diff
@@ -0,0 +1,64 @@
+            lhs.errorMessage != rhs.errorMessage {
+            return false
+        }
+        
```
Missing a check for the error message. Also, you have all of these checks be `guard else`.

Nitewriter commented on NameGame/NameGame/UI/UserGrid/ViewController/UserGridViewController.swift:
```diff
@@ -0,0 +1,168 @@
+        lifecycleObserver?.viewDidDisappear(animated)
+    }
+
+    override func didReceiveMemoryWarning() {
```
Let's remove this function for now since we are not using it.

Nitewriter commented on NameGame/NameGame/UI/UserGrid/ViewController/UserGridViewController.swift:
```diff
@@ -0,0 +1,168 @@
+    override func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        let cell = UserRecordCell.dequeueCell(in: collectionView, with: UserGridViewController.reuseCellIdentifier, for: indexPath)
+        
+        let cachedImage = imageCache.image(for: viewModel.userRecords![indexPath.item])
```
Let's use the nil coalescing operator so we don't need to force unwrap the user records.

```swift
let userRecords = viewModel.userRecords ?? []
...
userRecords[indexPath.item]
```

Nitewriter commented on NameGame/NameGame/UI/UserGrid/ViewController/UserGridViewController.swift:
```diff
@@ -0,0 +1,168 @@
+            activityIndicator.stopAnimating()
+        }
+        
+        guard let errorMessage = viewModel.errorMessage else {
```
Let's move this `guard` to the `presentAlert` function and change it's interface.

```swift
func presentAlert(with viewModel) {
   guard let errorMessage = ...
}
```

Then we can just call it from the `newState` function.

```swift
viewModel = state
renderLoadingStatus(with: viewModel)
renderUserRecords(with: viewModel) // <-- We are missing this one. 😉
presentAlert(with: viewModel)
```

Nitewriter reviewed (changes requested):
Really close, only other feedback I would give you is to make any properties `private` or `fileprivate` if they do not need to be accessed or changed by an outside party.

Commit:
All: made properties private or fileprivate if possible
MenuWorkflowPresentationEventObserver protocol: edited function names to present*
UserGridViewModel: guarded on all properties when implementing the equatable protocol
UserGridViewController: removed didRecieveMemoryWarning, added renderUserRecords, updated presentAlert

Nitewriter reviewed (approved):
LGTM! 💰
